### PR TITLE
Continuation PR for #256: Implement std::error::Error for errors

### LINKED ISSUE #247: PasrseError and error Trait
Hi,

I was wondering why the ParseError struct does not implement the [std::error::Error](https://doc.rust-lang.org/stable/std/error/trait.Error.html) trait? I was hoping to wrap it in a custom error enum type and use the "?" operator when calling parse functions. From what I've tried so far has not allowed me to do so.

### PR DESCRIPTION
**Pull request by dutchmartin**
This PR is the continuation of https://github.com/servo/rust-cssparser/pull/256.
Some changes are due to me running cargo `fmt`
I have implemented [description](https://doc.rust-lang.org/beta/std/error/trait.Error.html#method.description), but I can not satisfy the borrow checker with a reference of a String that will get deleted after the function ends. ([see image](https://user-images.githubusercontent.com/5509155/67152136-17080900-f2d1-11e9-9c4f-a62829a0a450.png)). Therefore I used a static string pointing the user in the right direction.

Closes #256 
Fixes #247

### CODE REVIEW

**Comment by nox** [comment]
>  I have not implemented [description](https://doc.rust-lang.org/beta/std/error/trait.Error.html#method.description) because it is deprecated and i can not satisfy the borrow checker with a reference of a String that will get deleted after the function ends. ([see image](https://user-images.githubusercontent.com/5509155/67152136-17080900-f2d1-11e9-9c4f-a62829a0a450.png))

Please implement `description` as I already asked in #256.

**Comment by dutchmartin** [comment]
> > I have not implemented [description](https://doc.rust-lang.org/beta/std/error/trait.Error.html#method.description) because it is deprecated and i can not satisfy the borrow checker with a reference of a String that will get deleted after the function ends. ([see image](https://user-images.githubusercontent.com/5509155/67152136-17080900-f2d1-11e9-9c4f-a62829a0a450.png))
> 
> Please implement `description` as I already asked in #256.

Done, please let me know what you think about proposed solution.

**Comment by dutchmartin** [comment]
@nox @SimonSapin Does one of you have time to review this?

File: procedural-masquerade/lib.rs [hunk 1]
```diff
@@ -244,33 +244,33 @@ macro_rules! define_invoke_proc_macro {
...
-                                            //
-                                            // Later, `stringify!(...)` is expanded into a string literal
-                                            // which is then ignored.
-                                            // Using `stringify!` enables passing arbitrary tokens
-                                            // rather than only what can be parsed as a const expression.
-                                            Input = (0, stringify! $paren ).0
-                                        }
-                                    }
-                                }
+                                                    ($proc_macro_name: ident ! $paren: tt) => {
```
Comment (emilio) [inline procedural-masquerade/lib.rs]: This seems a rustfmt bug that should be fixed and undone.

File: src/tokenizer.rs [hunk 2]
```diff
@@ -191,6 +191,68 @@ impl<'a> Token<'a> {
...
+            Ident(name) => format!("A ident token '{}'", name),
+            AtKeyword(value) => format!("The value '{}' does not include the `@` marker", value),
+            Hash(value) => format!("The value '{}' does not include the `#` marker", value),
+            IDHash(value) => format!(
+                "The value '{}' does not include the `#` marker, but has a valid ID selector",
+                value
+            ),
+            QuotedString(value) => format!("The value '{}' does not include the quotes", value),
+            UnquotedUrl(value) => format!(
+                "The value '{}' does not include the `url(` `)` markers.",
```
Comment (emilio) [inline src/tokenizer.rs]: This doesn't look right... `UnquotedUrl` is just `url(string-without-quotes)`, not any kind of error.

File: src/tokenizer.rs [hunk 3]
```diff
@@ -191,6 +191,68 @@ impl<'a> Token<'a> {
...
+    pub(crate) fn description(&self) -> String {
+        match self {
+            Ident(name) => format!("A ident token '{}'", name),
+            AtKeyword(value) => format!("The value '{}' does not include the `@` marker", value),
+            Hash(value) => format!("The value '{}' does not include the `#` marker", value),
+            IDHash(value) => format!(
+                "The value '{}' does not include the `#` marker, but has a valid ID selector",
+                value
+            ),
+            QuotedString(value) => format!("The value '{}' does not include the quotes", value),
```
Comment (emilio) [inline src/tokenizer.rs]: Same for this, I don't know what this is about?

File: src/parser.rs [hunk 4]
```diff
@@ -53,6 +55,20 @@ pub enum BasicParseErrorKind<'i> {
     QualifiedRuleInvalid,
 }
 
+impl<'i> BasicParseErrorKind<'i> {
+    fn description(&self) -> String {
+        match self {
+            BasicParseErrorKind::UnexpectedToken(token) => {
+                format!("Unexpected token: '{}'", token.description())
```
Comment (emilio) [inline src/parser.rs]: I would just do: `format!("Unexpected token: {:?}", token)`, and remove `Token::description()`.

File: src/tokenizer.rs [hunk 5]
```diff
@@ -191,6 +191,68 @@ impl<'a> Token<'a> {
...
+            Semicolon => String::from(";"),
+            Comma => String::from(","),
+            IncludeMatch => String::from("~="),
+            DashMatch => String::from("|="),
+            PrefixMatch => String::from("^="),
+            SuffixMatch => String::from("$="),
+            SubstringMatch => String::from("*="),
+            CDO => String::from("<!--"),
+            CDC => String::from("-->"),
+            Function(name) => format!("The value ({}) does not include the `(` marker", name),
```
Comment (emilio) [inline src/tokenizer.rs]: To be clear, these are legit token types. If the input is `foo()` and I do `input.expect_ident()`, I get an unexpected token error with the function, but there's nothing the "doesn't include the `(` marker" or anything like that.

**Comment by jdm** [comment]
Want to review the new commit @emilio?

File: src/parser.rs [hunk 6]
```diff
@@ -53,6 +55,20 @@ pub enum BasicParseErrorKind<'i> {
     QualifiedRuleInvalid,
 }
 
+impl<'i> BasicParseErrorKind<'i> {
+    fn description(&self) -> String {
+        match self {
```
Comment (emilio) [inline src/parser.rs:60]: Can we move this to `Display::fmt`?

File: src/parser.rs [hunk 7]
```diff
@@ -62,6 +78,18 @@ pub struct BasicParseError<'i> {
     pub location: SourceLocation,
 }
 
+impl<'i> fmt::Display for BasicParseError<'i> {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str(self.kind.description().as_str())
```
Comment (emilio) [inline src/parser.rs:83]: That way this doesn't need the extra allocation.

File: src/parser.rs [hunk 8]
```diff
@@ -156,6 +184,27 @@ impl<'i, T> ParseError<'i, T> {
     }
 }
 
+impl<'i, T> fmt::Display for ParseError<'i, T>
+where
+    T: fmt::Display,
+{
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        match &self.kind {
+            ParseErrorKind::Basic(basic_kind) => formatter.write_str(&basic_kind.description()),
```
Comment (emilio) [inline src/parser.rs:193]: This can just use `basic_kind.fmt(formatter)`

File: src/parser.rs (hunk 7 above)
Comment (emilio) [reply src/parser.rs:83]: Also, it may be nice to expose self.location, but that may be fine as a follow-up.

**Comment by categulario** [comment]
Is there a way I can help with this?

**Comment by dutchmartin** [comment]
@categulario If you would like to review and or make changes, please do. I wanted to finish this, but did not manage to. You have push access to my fork.
BTW: the concerns addressed by @emilio are reasonable.

File: src/parser.rs (hunk 7 above)
Comment (categulario) [reply src/parser.rs]: Done in c962d4a

**Comment by categulario** [comment]
Thanks @dutchmartin . I addressed @emilio 's concerns in c962d4a .

**Pull request closed without merging**